File: 01_GOF23/DP_examples/Python/02_Structural/Decorator/decorator.py
from component import IShape, Color

class Decorator(IShape):
    def __init__(self, shape):
        self.shape_component = shape
    
    def draw(self):
        if self.shape_component:
            self.shape_component.draw()
    
    def fill(self, color: Color):
        if self.shape_component:
            self.shape_component.fill(color)

class ShapeDecorator(Decorator):
    def __init__(self, shape : IShape):
        super().__init__(shape)
    
    def _before_draw(self):
        print("Before Draw() >> Do...")
    
    def _after_draw(self):
        print("When Draw Complete >> Do...")
    
    def _before_fill(self):
        print("Before Fill() >> Do...")
    
    def _after_fill(self):
        print("When Fill Complete >> Do...")
    
    def draw(self):
        self._before_draw()
        super().draw()
        self._after_draw()
    
    def fill(self, color: Color):
        self._before_fill()     
        super().fill(color)
        self._after_fill()

File: 01_GOF23/DP_examples/Rust/02_Structural/Decorator/decorator.rs
use crate::component::{Color, IShape};

pub struct Decorator<T: IShape> {
    shape: T,
}

impl<T: IShape> Decorator<T> {
    pub fn new(shape: T) -> Self {
        Decorator { shape }
    }
}

impl<T: IShape> IShape for Decorator<T> {
    fn draw(&self) {
        self.shape.draw();
    }
    
    fn fill(&self, color: Color) {
        self.shape.fill(color);
    }
}

pub struct ShapeDecorator<T: IShape> {
    decorator: Decorator<T>,
}

impl<T: IShape> ShapeDecorator<T> {
    pub fn new(shape: T) -> Self {
        ShapeDecorator {
            decorator: Decorator::new(shape),
        }
    }
    
    fn before_draw(&self) {
        println!("Before Draw() >> Do...");
    }
    
    fn after_draw(&self) {
        println!("When Draw Complete >> Do...");
    }
    
    fn before_fill(&self) {
        println!("Before Fill() >> Do...");
    }
    
    fn after_fill(&self) {
        println!("When Fill Complete >> Do...");
    }
}

impl<T: IShape> IShape for ShapeDecorator<T> {
    fn draw(&self) {
        self.before_draw();
        self.decorator.draw();
        self.after_draw();
    }
    
    fn fill(&self, color: Color) {
        self.before_fill();
        self.decorator.fill(color);
        self.after_fill();
    }
}
